src/src/com/microsoft/aad/adal/TokenCacheAccessor.java

```diff
@@ -0,0 +1,271 @@
+        }
+        
+        if (StringExtensions.IsNullOrBlank(authority)) {
+            throw new IllegalArgumentException("authority");
+        }
+        
+        mTokenCacheStore = tokenCacheStore;
+        mAuthority = authority;
+    }
+    
+    /**
+     * @return {@link TokenCacheItem} for regular token cache entry.  
+     */
+    TokenCacheItem getRegularTokenCacheItemWithAT(final String resource, final String clientId, final String user) {
+        final String cacheKey = CacheKey.createCacheKeyForRTEntry(mAuthority, resource, clientId, user);
+        return mTokenCacheStore.getItem(cacheKey);
+    }
+    
+    /**
+     * @return {@link TokenCacheItem} for MRRT token cache entry.  
+     */
+    TokenCacheItem getMRRTItem(final String clientId, final String user) {
+        final String cacheKey = CacheKey.createCacheKeyForMRRT(mAuthority, clientId, user);
+        return mTokenCacheStore.getItem(cacheKey);
+    }
+    
+    /**
+     * @return {@link TokenCacheItem} for FRT token cache entry.  
+     */
+    TokenCacheItem getFRTItem(final String familyClientId, final String user) {
+        if (StringExtensions.IsNullOrBlank(user)) {
+            return null;
+        }
+        
+        final String cacheKey = CacheKey.createCacheKeyForFRT(mAuthority, familyClientId, user);
+        return mTokenCacheStore.getItem(cacheKey);
+    }
+    
+    /**
+     * Remove existing token cache item if needed and update token cache with returned auth result. 
+     */
+    void updateCachedItemToResult(final String resource, final String clientId, final AuthenticationResult result, 
+            final TokenCacheItem cachedItem) {
+        if (result == null) {
+            return;
+        }
+        
+        // remove Item if oauth2_error is invalid_grant
+        if (AuthenticationConstants.OAuth2ErrorCode.INVALID_GRANT.equalsIgnoreCase(result.getErrorCode())) {
+            Logger.v(TAG, "Received INVALID_GRANT error code, remove existing cache entry.");
+            removeTokenCacheItem(cachedItem, resource);
+        } else {
+            Logger.v(TAG, "Save returned AuthenticationResult into cache.");
+            if (cachedItem != null && cachedItem.getUserInfo() != null && result.getUserInfo() == null) {
+                result.setUserInfo(cachedItem.getUserInfo());
+                result.setIdToken(cachedItem.getRawIdToken());
+                result.setTenantId(cachedItem.getTenantId());
+            }
+            
+            updateTokenCache(resource, clientId, result);
+        }
+    }
+    
+    /**
+     * Udpate token cache with returned auth result. 
+     * @param result
+     */
+    void updateTokenCache(final String resource, final String clientId, final AuthenticationResult result) {
+        if (result == null || StringExtensions.IsNullOrBlank(result.getAccessToken())) {
+            return;
+        }
+        
+        //@note: The current way for setting token into cache is
```

Comment:
Done. 

---
In reply to: [62961067](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/611#discussion_r62961067) [](ancestors = 62961067,62917602)